shelley-ma/shelley-ma-test/src/Test/Cardano/Ledger/ShelleyMA/Serialisation/Generators.hs

```diff
@@ -0,0 +1,225 @@
+import Test.Shelley.Spec.Ledger.Serialisation.EraIndepGenerators
+  ( EraGen (..),
+    maxTxWits,
+  )
+import Test.Tasty.QuickCheck (Gen, choose, elements)
+
+{-------------------------------------------------------------------------------
+  ShelleyMAEra Generators
+
+  Generators used for roundtrip tests, generated values are not
+  necessarily valid
+-------------------------------------------------------------------------------}
+
+genTxBody ::
+  ( ShelleyBased era,
+    Abstract.TxBody era ~ MA.TxBody era,
+    Mock (Crypto era),
+    Arbitrary (Abstract.Value era)
+  ) =>
+  Gen (MA.TxBody era)
+genTxBody =
+  MA.TxBody
+    <$> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+    <*> arbitrary
+
+genTx ::
+  ( ShelleyBased era,
+    Abstract.TxBody era ~ MA.TxBody era,
+    Mock (Crypto era),
+    Arbitrary (WitnessSet era),
+    Arbitrary (Abstract.Value era)
+  ) =>
+  Gen (Tx era)
+genTx =
+  Tx
+    <$> genTxBody
+    <*> (resize maxTxWits arbitrary)
+    <*> arbitrary
+
+genBlock ::
+  forall era.
+  ( ShelleyBased era,
+    Abstract.TxBody era ~ MA.TxBody era,
+    Mock (Crypto era),
+    Arbitrary (WitnessSet era),
+    Arbitrary (Abstract.Value era)
+  ) =>
+  Gen (Block era)
+genBlock = do
+  let KeySpace_ {ksCoreNodes} = geKeySpace (genEnv p)
+  prevHash <- arbitrary :: Gen (HashHeader (Crypto era))
+  allPoolKeys <- elements (map snd ksCoreNodes)
+  txs <- listOf genTx
+  curSlotNo <- SlotNo <$> choose (0, 10)
+  curBlockNo <- BlockNo <$> choose (0, 100)
+  epochNonce <- arbitrary :: Gen Nonce
+  let kesPeriod = 1
+      keyRegKesPeriod = 1
+      ocert = mkOCert allPoolKeys 1 (KESPeriod kesPeriod)
+  return $
+    mkBlock
+      prevHash
+      allPoolKeys
+      txs
+      curSlotNo
+      curBlockNo
+      epochNonce
+      kesPeriod
+      keyRegKesPeriod
+      ocert
+  where
+    p :: Proxy era
+    p = Proxy
+
+genMAScript ::
+  (Era era, Mock (Crypto era)) =>
+  Gen (MA.Script era)
+genMAScript =
+  oneof
+    [ MA.ScriptMSig <$> arbitrary,
+      -- TODO @uroboros Write recursive (and sized) generator that engages all Timelock contructors
+      MA.ScriptTimelock . Interval <$> arbitrary
+    ]
+
+genWitnessSet ::
+  forall era.
+  ( Era era,
+    Mock (Crypto era),
+    ValidateScript era,
+    FromCBOR (Annotator (Abstract.Script era)),
+    EraGen era
+  ) =>
+  Gen (WitnessSet era)
+genWitnessSet =
+  WitnessSet
+    <$> arbitrary
+    <*> (mscriptsToWits <$> (vectorOf 5 (genScript @era)))
+    <*> arbitrary
+  where
+    mscriptsToWits = Map.fromList . map (\s -> (hashScript s, s))
+
+{-------------------------------------------------------------------------------
+  MaryEra Arbitrary Instances
+-------------------------------------------------------------------------------}
+
+instance
+  (Mock c) =>
+  EraGen (MaryEra c)
+  where
+  genScript = genMAScript
+
+instance Arbitrary (MA.TxBody (MaryEra C_Crypto)) where
```

Comment:
Why are these all fixed to `C_Crypto`? (Same for Allegra) Consensus can't use `C_Crypto`.

Fortunately, you can generalise these all to `Mock c => .. c`, I have tried it out myself :slightly_smiling_face:.

Reply:
sure, I'll go with `Mock c` - actually started with that and then specialised ;/